docs/design/settings-and-sessions.rst

```diff
@@ -0,0 +1,132 @@
+* Should we break out the virtual environment settings (venv location, baseline packages) into its own file? [+1]
+* What levels of config do we need? Defaults? One/multiple settings files? Override at instance level?
+* Do we still need to look in the application directory as well as the data directory? [-0]
+
+Exit Handlers
+~~~~~~~~~~~~~
+
+Registered exit handlers to ensure that files are saved when Mu exits. (This
+could probably be alternatively achieved within the Qt app). The advantage of
+this is that the save is automatic; the disadvantage is that it's a little
+hidden.
+
+Not currently writing to disc as soon as updated: having an exit handler ensures
+the settings will be written, even in the event of an unhandled exception.
+And it's not clear what advantage an "autosave" would offer.
+
+
+Levels of Config
+~~~~~~~~~~~~~~~~
+
+Allowing three levels of data: the defaults for each setting type, held in
+a class dictionary; possible overrides at class instantiation [I'm not clear
+where this would be used; it can probably go]; and the .json files.
+
+The ``load`` function merges into the existing settings. Most commonly this means
+it'll be preceded by a call to ``reset``. But it could be used to implement a
+cascade of settings, eg where an admin sets site-wide settings which are then
+overridden by user settings.
+
+Amnesia / Read-only / Reset modes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To support the possible "modes" above -- amnesia, read-only etc. there is a
+``readonly`` flag on each settings object, preventing it from being written to
+disc; and a ``reset`` method which will return to default settings. This last
+can be used either to "forget" any loaded or set settings; or before reloading
+from a different file.
+
+So *amnesia mode* is implemented by calling ``reset`` without ``load`` and settings ``readonly``.
+*Read-only mode* is implemented by calling ``reset`` followed by ``load`` and setting ``readonly``
+And *reset mode* is implemented by calling ``reset`` without ``load`` and *not* setting ``readonly``
+
+Failure modes
+~~~~~~~~~~~~~
+
+It's critical that we should recover well from not being able to read or to
+write settings files, whether that's a file system failure or invalid JSON.
+Regardless of the approach we should definitely log any exception, or log a
+warning where there's no exception as such but, say, a missing file.
+
+Reading
++++++++
+
+* A failure to find/open a settings file is considered usual: it's expected
+  that, the first time around, a user settings file won't exist to be read.
+  The loader will log a warning and carry on as though it had found it empty
+* A failure to read the JSON from a settings file is more complicated. For
+  pragmatic purposes, the intention is here is: log a warning; quarantine the
+  file; and carry on as though it had been found empty. That way the editor
+  continues to work, albeit in "reset" mode, and the failing file is available
+  for debugging.
```

Comment:
Although not something to be done by this module itself, I think we should show an error to the user.
The most likely situation of a malformed JSON file is the user trying to change a setting, and in that case loading the defaults without immediate visible warning can be confusing and/or misleading.

Reply:
@carlosperate agreed, but that's part of a wider piece: how to keep the user informed during startup. I have a notion about essentially extending the current "Splash" frame to have a status line with a slot which startup processes can signal, indicating progress or errors / failures etc.

Of course, we might have a failure even to start up a UI at all. But I regard that as a slightly different problem